pkg/fs/fs-bucket-listobjects.go

```diff
@@ -20,18 +20,155 @@ import (
 	"fmt"
 	"net/url"
 	"os"
-	"path"
 	"path/filepath"
 	"strings"
+	"time"
 
 	"github.com/minio/minio/pkg/probe"
 )
 
+const (
+	// listObjectsLimit - maximum list objects limit.
+	listObjectsLimit = 1000
+)
+
+// ObjectInfo - object info.
+type ObjectInfo struct {
+	Bucket       string
+	Name         string
+	ModifiedTime time.Time
+	ContentType  string
+	MD5Sum       string
+	Size         int64
+	IsDir        bool
+}
+
+type treeWalkResult struct {
+	objectInfo ObjectInfo
+	err        error
+	end        bool
+}
+
+type treeWalker struct {
+	ch       <-chan treeWalkResult
+	timedOut bool
+}
+
+func treeWalk(bucketDir, prefixDir, entryPrefixMatch, marker string, recursive bool, send func(treeWalkResult) bool, count *int) bool {
```

Comment:
If you wanted to change the current behaviour of marker handling, you could add filter function to readDirAll() to handle it.  That will be much cleaner code.